Add Scheduler (#143)
PaperMC/docs: docs/paper/dev/api/scheduler.md

﻿---
slug: /dev/scheduler
---

> # Task Scheduler

Reviewer: Why is the page called task scheduler if you immediately refer to it with the `BukkitScheduler` name. Maybe just name the page "The Bukkit Scheduler" or something along those lines?

Author: I'd prefer something like `Scheduling tasks` over including Bukkit in the name

> The `BukkitScheduler` can be used to schedule your code to be run later or run it repeatedly.

Reviewer: ```suggestion
The `BukkitScheduler` can be used to schedule your code to be run later or repeatedly.
```

Author: Will change

a single execution of that loop is called a 'tick'.

In Minecraft's case the amount of ticks per second is 20, or one tick every 50 milliseconds,
meaning that the game loop is executed 20 times per second. A tick taking more than 50ms to execute is the moment
when your server starts to fall behind on its work and lag.

### Converting between human units and Minecraft ticks

Every method of the scheduler that takes a delay or period uses ticks as a unit of time.

Converting from human units to ticks and back is as simple as:  
`ticks = seconds * 20`  
`seconds = ticks / 20`

You can make your code more readable by using the
[`TimeUnit`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/TimeUnit.html)
enum, e.g. to convert 5 minutes to ticks and back:  
`TimeUnit.MINUTES.toSeconds(5) * 20`  
`TimeUnit.SECONDS.toMinutes(ticks / 20)`

## Obtaining the scheduler

To obtain a scheduler you can use the instance method on the `Server` class, e.g. in your `onEnable` method:

```java
@Override
public void onEnable() {
	BukkitScheduler scheduler = this.getServer().getScheduler();
}
```

## Scheduling tasks

Scheduling a task requires you to pass:

> - your plugin's instance,
> - the code to run, either with a `Runnable` or `Consumer<BukkitTask>`,
> - the delay in ticks before the task should run,
> - if you're scheduling a repeating task - the period in ticks between each execution of the task.

Reviewer: ```suggestion
- Your plugin's instance,
- The code to run, either with a `Runnable` or `Consumer<BukkitTask>`,
- The delay in ticks before the task should run for the first time,
- If you're scheduling a repeating task - the period in ticks between each execution of the task.
```

Author: Think of it as a long sentence, hence the commas and the period at the end,
will add `for the first time`

and will be affected if the server is lagging, an example would be 20 ticks not being exactly 1 second.

If you need a scheduler that runs independently of the server consider using your own
[`ScheduledExecutorService`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/ScheduledExecutorService.html).
You can follow [this guide](https://www.baeldung.com/java-executor-service-tutorial#ScheduledExecutorService) to learn how to use it.

:::

### Difference between `Runnable` and `Consumer<BukkitTask>`

#### Using `Runnable`

The `Runnable` interface is used for the simplest tasks that don't require a `BukkitTask` instance.

You can either implement it in a separate class, e.g.:

```java
public class MyRunnableTask implements Runnable {

	private final MyPlugin plugin;
	
	public MyRunnableTask(MyPlugin plugin) {
		this.plugin = plugin;
	}
	
	@Override
	public void run() {
		this.plugin.getServer().broadcast(Component.text("Hello, World!"));
	}

}
```
```java
scheduler.runTaskLater(plugin, new MyRunnableTask(plugin), 20);
```

Or use a lambda expression which is great for simple and short tasks:

```java
scheduler.runTaskLater(

> 	plugin, 
> 	// Lambda:
> 	() -> {
> 		this.plugin.getServer().broadcast(Component.text("Hello, World!"));
> 	},
> 	// End of the lambda
> 	20);

Reviewer: ```suggestion
	plugin, /* Lambda: */ () -> {
		this.plugin.getServer().broadcast(Component.text("Hello, World!"));
	}, /* End of the lambda */ 20);
```

Keeps it styled the same as the next example and looks cleaner

Author: Will change

```

#### Using `Consumer<BukkitTask>` {#using-consumerbukkittask}

The `Consumer` interface is used for tasks that require a `BukkitTask` instance (usually in repeated tasks),
e.g. when you want to cancel the task from inside it.

You can either implement it in a separate class similarly to the `Runnable`, e.g.:

```java
public class MyConsumerTask implements Consumer<BukkitTask> {
	
	private final LivingEntity entity;
	
	public MyConsumerTask(LivingEntity entity) {
		this.entity = entity;
	}
	
	@Override
	public void accept(BukkitTask task) {
		if(this.entity.isDead()) {
			task.cancel(); // The entity died, there's no point
			return;        // in running the code anymore.
		}
		
		this.entity.addPotionEffect(new PotionEffect(PotionEffectType.SPEED, 20, 1));
	}
	
}
```
```java
scheduler.runTaskTimer(plugin, new MyConsumerTask(someEntity), 0, 20);
```

Or use a lambda expression which again is much cleaner for short and simple tasks:

```java
scheduler.runTaskTimer(plugin, /* Lambda: */ task -> {

> 	if(this.entity.isDead()) {

Reviewer 2: Missing a space between `if` and `(`

﻿---
slug: /dev/scheduler
---

# Task Scheduler

The `BukkitScheduler` can be used to schedule your code to be run later or run it repeatedly.

## What is a tick?

Every game runs something called a game loop which essentially executes all the logic of the game over and over,
a single execution of that loop is called a 'tick'.

> In Minecraft's case the amount of ticks per second is 20, meaning that the game loop is executed 20 times per second.
> With some math we can see that 1 tick is equal to 50 milliseconds. A tick taking more than 50ms to execute is the moment

Reviewer 3: ```suggestion
In Minecraft's case the amount of ticks per second is 20, or one tick every 50 milliseconds,
meaning that the game loop is executed 20 times per second. A tick taking more than 50ms to execute is the moment
```

If you need a scheduler that runs independently of the server consider using your own
[`ScheduledExecutorService`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/ScheduledExecutorService.html).
You can follow [this guide](https://www.baeldung.com/java-executor-service-tutorial#ScheduledExecutorService) to learn how to use it.

:::

### Difference between `Runnable` and `Consumer<BukkitTask>`

#### Using `Runnable`

The `Runnable` interface is used for the simplest tasks that don't require a `BukkitTask` instance.

You can either implement it in a separate class, e.g.:

```java
public class MyRunnableTask implements Runnable {

	private final MyPlugin plugin;
	
	public MyRunnableTask(MyPlugin plugin) {
		this.plugin = plugin;
	}
	
	@Override
	public void run() {
		this.plugin.getServer().broadcast(Component.text("Hello, World!"));
	}

}
```
```java
scheduler.runTaskLater(plugin, new MyRunnableTask(plugin), 20);
```

Or use a lambda expression which is great for simple and short tasks:

```java
scheduler.runTaskLater(
	plugin, 
	// Lambda:
	() -> {

> 		this.plugin.getServer().broadcast(Component.text("Hello, World!")

Reviewer 4: Missing a closing `)`

```java
scheduler.runTaskTimer(plugin, /* Lambda: */ task -> {
	if(this.entity.isDead()) task.cancel();
	this.entity.addPotionEffect(new PotionEffect(PotionEffectType.SPEED, 20, 1));
} /* End of the lambda */, 0, 20);
```

## Examples of tasks on the main thread

### A task to run later once

This task will run a single time after the delay specified in ticks, in this case 1 second.

```java
scheduler.runTaskLater(plugin, () -> {
	server.broadcast(Component.text("Hello, World!"));
}, 20);
```

### A task to run later repeatedly

This task will run repeatedly, first time after the delay specified in ticks - in this case 1 second -
and then every period specified in ticks - in this case 5 seconds.

```java
scheduler.runTaskTimer(plugin, () -> {
	server.broadcast(Component.text("Hello, World!"));
}, 20, 5 * 20);
```

### A repeating task to be canceled later

This task will run repeatedly, first time after 10 seconds, and then every 5 seconds. After 10 minutes
the task will be canceled entirely.

```java
BukkitTask task = scheduler.runTaskTimer(plugin, () -> {
	server.broadcast(Component.text("Hello, World!"));
}, 10 * 20, 5 * 20);

> scheduler.runTaskLater(plugin, () -> task.cancel(), TimeUnit.MINUTES.toSeconds(10) * 20);

Author: Not sure that's what you meant but changed

﻿---
slug: /dev/scheduler
---

# Scheduler

The `BukkitScheduler` can be used to schedule your code to be run later or run it repeatedly.

## What is a tick?

Every game runs something called a game loop which essentially executes all the logic of the game over and over,
a single execution of that loop is called a 'tick'.

In Minecraft's case the amount of ticks per second is 20, meaning that the game loop is executed 20 times per second.
With some math we can see that 1 tick is equal to 50 milliseconds. A tick taking more than 50ms to execute is the moment
when your server starts to fall behind on its work and lag.

> ### Converting between SI units and Minecraft ticks

Author: Missed when changing to human units as below
>Converting from human units to ticks and back is as simple as:  
agreed & changed

Every method of the scheduler that takes a delay or period uses ticks as a unit of time.

> :::warning

Reviewer: I don’t think this warrants a warning. Maybe just info

Author: IMO time units for async methods would be somewhat expected to be milliseconds (when there's no `TimeUnit` param), however they are not therefore the big red warning

The above also applies to the asynchronous methods. If you need a scheduler with millisecond precision, create your own
[`ScheduledExecutorService`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/ScheduledExecutorService.html).

:::

Converting from human units to ticks and back is as simple as:  

> `ticks = seconds * 20`  

Author: I don't see how secondsPassed is better, imo readable as is

> `ticks = seconds * 20`  

Reviewer 4: I think its fine as is.

`seconds = ticks / 20`

:::note

> In Java and Kotlin to not lose precision you must use `ticks / 20.0`.

Reviewer: I don’t think this is the place to explain basic java

Author: considering the IDE does display a warning when the variable's type is `float`/`double` and above is pseudo-code, agreed

:::

You can make your code more readable by using the
[`TimeUnit`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/TimeUnit.html)
enum, e.g. to convert 5 minutes to ticks and back:  
`TimeUnit.MINUTES.toSeconds(5) * 20`  
`TimeUnit.SECONDS.toMinutes(ticks / 20)`

## Obtaining the scheduler

To obtain a scheduler you can use the instance method on the `Server` class, e.g. in your `onEnable` method:

```java
@Override
public void onEnable() {

> 	BukkitScheduler scheduler = this.getServer().getScheduler();

Author: convention about `this` was already broken with my previous PR of Custom InventoryHolders, will leave for staff to decide

public void onEnable() {
	BukkitScheduler scheduler = this.getServer().getScheduler();
}
```

## Scheduling tasks

Scheduling a task requires you to pass:
- your plugin's instance,
- the code to run, either with a `Runnable` or `Consumer<BukkitTask>` (the differences and usage are explained below),
- the delay in ticks before the task should run,
- the period in ticks between each execution of the task if you're scheduling a repeating task.

> ### Using `Runnable`

Author: not so glad of the structure / naming of this section

	
	@Override
	public void run() {
		try {
			this.plugin.getServer().broadcast(Component.text("Hello, World!"));
		} finally {
			this.plugin.getServer().getScheduler().runTaskLater(
				this.plugin,
				this,
				this.random.nextInt(20, 101)
			);
		}
	}
	
}
```

```java
scheduler.runTaskLater(plugin, new MyDynamicTask(plugin), 10 * 20);
```

:::note

We schedule the task again in the `run` method, that's why to schedule the task initially
we use the `runTaskLater` method too and not `runTaskTimer`.

:::

:::info

We use a try-finally block to ensure the task is scheduled again. Otherwise if the methods before scheduling
the tasks would throw an exception the task would never be scheduled again.

> :::

Author: not sure which warrants a note and which an info? opinions very much welcome

Author: guess note reiterates sth that can be missed and info explains stuff 🤷🏻
